src/Compilers/CSharp/Portable/Symbols/ReferenceManager.cs

```diff
@@ -306,183 +306,163 @@ private void InitializeAssemblyReuseData(AssemblySymbol assemblySymbol, Immutabl
-
-                var resolutionDiagnostics = default(ImmutableArray<Diagnostic>);
-                var referenceMap = default(ImmutableArray<ResolvedReference>);
-                var diagnostics = DiagnosticBag.GetInstance();
+                var resolutionDiagnostics = DiagnosticBag.GetInstance();
+                var implicitlyResolvedReferences = ArrayBuilder<MetadataReference>.GetInstance();
 
                 try
                 {
-                    referenceMap = ResolveMetadataReferences(
+                    IDictionary<string, MetadataReference> boundReferenceDirectiveMap;
+                    ImmutableArray<MetadataReference> boundReferenceDirectives;
+                    ImmutableArray<AssemblyData> referencedAssemblies;
+                    ImmutableArray<PEModule> modules; // To make sure the modules are not collected ahead of time.
+                    ImmutableArray<MetadataReference> references;
+
+                    ImmutableArray<ResolvedReference> referenceMap = ResolveMetadataReferences(
                         compilation,
                         out references,
                         out boundReferenceDirectiveMap,
                         out boundReferenceDirectives,
                         out referencedAssemblies,
                         out modules,
-                        diagnostics);
-
-                    resolutionDiagnostics = diagnostics.ToReadOnly();
-                }
-                finally
-                {
-                    diagnostics.Free();
-                }
-
-                var assemblyBeingBuiltData = new AssemblyDataForAssemblyBeingBuilt(new AssemblyIdentity(name: SimpleAssemblyName), referencedAssemblies, modules);
-
-                var assembliesArray = new AssemblyData[referencedAssemblies.Length + 1];
-                assembliesArray[0] = assemblyBeingBuiltData;
-                for (int i = 1; i < assembliesArray.Length; i++)
-                {
-                    assembliesArray[i] = referencedAssemblies[i - 1];
-                }
-
-                ImmutableArray<AssemblyData> allAssemblies = assembliesArray.AsImmutableOrNull();
+                        resolutionDiagnostics);
+
+                    var assemblyBeingBuiltData = new AssemblyDataForAssemblyBeingBuilt(new AssemblyIdentity(name: SimpleAssemblyName), referencedAssemblies, modules);
+                    var allAssemblyData = referencedAssemblies.Insert(0, assemblyBeingBuiltData);
+
+                    // Let's bind all the references and resolve missing one (if resolver is available)
+                    bool hasCircularReference;
+                    int corLibraryIndex;
+                    BoundInputAssembly[] bindingResult = Bind(
+                        ref allAssemblyData,
+                        compilation.Options.MetadataReferenceResolver,
+                        compilation.Options.MetadataImportOptions,
+                        implicitlyResolvedReferences,
```

Comment:
I used the same pattern as with diagnostics (the next argument), which is also a builder.